File: packages/llamaindex/src/retriever.ts
// ─────────────────────────────────────────────
// MemHQRetriever for LlamaIndex.js
//
// Implements the BaseRetriever `retrieve` contract: takes a query and
// returns an array of NodeWithScore-shaped results. Compatible with
// QueryEngine, RetrieverQueryEngine, RouterQueryEngine, agents, etc.
//
// We use a structural type rather than importing from `llamaindex` so
// this adapter keeps working across LlamaIndex's frequent API moves.
// ─────────────────────────────────────────────

import { MemoryClient } from "@memhq/sdk";

/** Structural shape compatible with LlamaIndex's NodeWithScore. */
export interface NodeWithScoreLike {
  node: {
    id_: string;
    text: string;
    metadata: Record<string, unknown>;
    getContent(): string;
  };
  score: number;
}

/** Structural shape of a LlamaIndex retriever query input. */
export interface RetrieveParams {
  query: string | { query: string };
}

export interface MemHQRetrieverOptions {
  apiKey?: string;
  userId: string;
  baseUrl?: string;
  limit?: number;
  mode?: "hybrid" | "vector" | "lexical";
  client?: MemoryClient;
}

export class MemHQRetriever {
  private readonly client: MemoryClient;
  private readonly userId: string;
  private readonly limit: number;
  private readonly mode: "hybrid" | "vector" | "lexical";

  constructor(opts: MemHQRetrieverOptions) {
    this.userId = opts.userId;
    this.limit = opts.limit ?? 10;
    this.mode = opts.mode ?? "hybrid";
    this.client =
      opts.client ??
      new MemoryClient({
        ...(opts.apiKey !== undefined ? { apiKey: opts.apiKey } : {}),
        ...(opts.baseUrl !== undefined ? { baseUrl: opts.baseUrl } : {}),
      });
  }

  async retrieve(params: RetrieveParams | string): Promise<NodeWithScoreLike[]> {
    const query =
      typeof params === "string"
        ? params
        : typeof params.query === "string"
          ? params.query
          : params.query.query;
    const res = await this.client.search({
      query,
      userId: this.userId,
      limit: this.limit,
      mode: this.mode,
    });
    return res.results.map<NodeWithScoreLike>((m) => ({
      node: {
        id_: m.id,
        text: m.content,
        metadata: {
          memory_id: m.id,
          confidence: m.confidence,
          type: m.type,
          source: "memhq",
        },
        getContent: () => m.content,
      },
      score: m.score,
    }));
  }
}

File: python/memhq-llamaindex/memhq_llamaindex/retriever.py
"""MemHQ-backed LlamaIndex retriever.

Implements ``llama_index.core.retrievers.BaseRetriever`` so it plugs
into ``RetrieverQueryEngine``, ``RouterQueryEngine``, agent tool
factories, etc.
"""

from __future__ import annotations

from typing import List, Optional

from llama_index.core.callbacks import CallbackManager
from llama_index.core.retrievers import BaseRetriever
from llama_index.core.schema import NodeWithScore, QueryBundle, TextNode

from memhq_llamaindex._client import _Transport, asearch, search


class MemHQRetriever(BaseRetriever):
    """Retriever backed by MemHQ ``/v1/memhq/search``."""

    def __init__(
        self,
        *,
        user_id: str,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        limit: int = 10,
        mode: str = "hybrid",
        callback_manager: Optional[CallbackManager] = None,
    ) -> None:
        super().__init__(callback_manager=callback_manager)
        self._transport = _Transport(api_key=api_key, base_url=base_url)
        self._user_id = user_id
        self._limit = limit
        self._mode = mode

    def _retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
        results = search(
            self._transport,
            query=query_bundle.query_str,
            user_id=self._user_id,
            limit=self._limit,
            mode=self._mode,
        )
        return [_to_node(r) for r in results]

    async def _aretrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
        results = await asearch(
            self._transport,
            query=query_bundle.query_str,
            user_id=self._user_id,
            limit=self._limit,
            mode=self._mode,
        )
        return [_to_node(r) for r in results]


def _to_node(r: dict) -> NodeWithScore:
    node = TextNode(
        id_=r.get("id", ""),
        text=r.get("content", ""),
        metadata={
            "memory_id": r.get("id"),
            "confidence": r.get("confidence"),
            "type": r.get("type"),
            "source": "memhq",
        },
    )
    return NodeWithScore(node=node, score=float(r.get("score", 0.0)))
